libs/utils/src/filter.c

```diff
@@ -17,421 +17,503 @@
+    celix_autofree char* any = NULL;
+    size_t anySize = 0;
+    celix_autoptr(FILE) stream = NULL;
+    int startPos = *pos;
+    while (filterString[*pos] != ')' && filterString[*pos] != '*') {
+        int rc;
+        if (filterString[*pos] == '\\') {
+            (*pos)++; // eat '\'
+        }
+        if (filterString[*pos] == '\0') {
+            celix_err_push("Filter Error: Unexpected end of string while parsing attribute value.");
+            return CELIX_INVALID_SYNTAX;
+        }
+        if (!stream) {
+            stream = open_memstream(&any, &anySize);
+            if (!stream) {
+                celix_err_push("Filter Error: Failed to open mem stream.");
+                return CELIX_ENOMEM;
             }
         }
+        rc = fputc(filterString[*pos], stream);
+        if (rc == EOF) {
+            celix_err_push("Filter Error: Failed to write to stream.\n");
+            return CELIX_FILE_IO_EXCEPTION;
+        }
+        (*pos)++;
     }
-    free(sub);
 
-    if (celix_arrayList_size(operands) == 0) {
-        fprintf(stderr, "Filter Error: Missing value.\n");
-        celix_arrayList_destroy(operands);
+    if (startPos == *pos) {
+        // empty any
+        *out = NULL;
+        return CELIX_SUCCESS;
+    }
+
+    int rc = fclose(celix_steal_ptr(stream));
+    if (rc != 0) {
+        celix_err_push("Filter Error: Failed to close stream.");
+        return CELIX_FILE_IO_EXCEPTION;
+    }
+
+    *out = celix_steal_ptr(any);
+    return CELIX_SUCCESS;
+}
+
+/**
+ * @brief Parses a substring filter.
+ * Returns a array list with at least 2 elements: The first element is the initial element, the last element is the
+ * final element. The initial and final element can be NULL.
+ *
+ * e.g.:
+ * - (foo=bar*) -> [bar, NULL]
+ * - (foo=*bar) -> [NULL, bar]
+ * - (foo=*bar*) -> [NULL, bar, NULL]
+ * - (foo=bar*bar) -> [bar, bar]
+ * - (foo=bar*bar*) -> [bar, bar, NULL]
+ */
+static celix_array_list_t* celix_filter_parseSubstring(const char* filterString, int* pos) {
+    celix_array_list_create_options_t ops = CELIX_EMPTY_ARRAY_LIST_CREATE_OPTIONS;
+    ops.simpleRemovedCallback = free;
+    celix_autoptr(celix_array_list_t) subs = celix_arrayList_createWithOptions(&ops);
+    if (!subs) {
+        celix_err_push("Filter Error: Failed to allocate memory.");
         return NULL;
     }
 
-    return operands;
+    celix_filter_skipWhiteSpace(filterString, pos);
+    if (filterString[*pos] == '*') {
+        // initial substring is NULL
+        // eat '*'
+        (*pos)++;
+        if(celix_arrayList_add(subs, NULL) != CELIX_SUCCESS) {
+            celix_err_push("Filter Error: Failed to add element to array list.");
+            return NULL;
+        }
+    }
+
+    char* element = NULL;
+    celix_status_t status;
+    do {
+        status = celix_filter_parseSubstringAny(filterString, pos, &element);
+        if (status != CELIX_SUCCESS) {
```

Comment:
This fix a possible double-free: a previous `element` may be added to the `arrayList` twice when `celix_filter_parseSubstringAny` returns error.